File: services/tradelog/tradelog-api/src/main/java/com/example/tradelog/api/repository/TradeSummaryModelRowMapper.java
package com.example.tradelog.api.repository;

import com.example.tradelog.api.spec.model.Action;
import com.example.tradelog.api.spec.model.TradeSummaryModel;
import com.example.tradelog.api.spec.model.TransactionType;
import org.springframework.jdbc.core.RowMapper;

import java.math.BigDecimal;
import java.sql.ResultSet;
import java.sql.SQLException;

public class TradeSummaryModelRowMapper implements RowMapper<TradeSummaryModel> {

    @Override
    public TradeSummaryModel mapRow(ResultSet rs, int rowNum) throws SQLException {
        TransactionType transactionType = TransactionType.Companion.lookup(rs.getString("transaction_type_fk"));

        if (TransactionType.SHARE.equals(transactionType)) {
            return fromStock(rs);
        } else if (TransactionType.OPTION.equals(transactionType)) {
            return fromOption(rs);
        } else if (TransactionType.DIVIDEND.equals(transactionType)) {
            return fromDividend(rs);
        }

        throw new IllegalArgumentException("Received unknown transaction type: " + transactionType);
    }

    private TradeSummaryModel fromStock(ResultSet rs) throws SQLException {
        Action action = Action.Companion.lookup(rs.getString("action_fk"));

        BigDecimal price = BigDecimal.valueOf(rs.getDouble("price"));

        if (Action.BUY.equals(action)) {
            price = price.negate();
        }

        BigDecimal fees = BigDecimal.valueOf(rs.getDouble("broker_fees"));
        BigDecimal quantity = BigDecimal.valueOf(rs.getInt("quantity"));

        return TradeSummaryModel.builder()
                .withSymbol(rs.getString("symbol"))
                .withTotal(price.multiply(quantity).subtract(fees))
                .withTrades(1)
                .build();
    }

    private TradeSummaryModel fromOption(ResultSet rs) throws SQLException {
        BigDecimal premium = BigDecimal.valueOf(rs.getDouble("premium"));
        BigDecimal fees = BigDecimal.valueOf(rs.getDouble("broker_fees"));
        BigDecimal quantity = BigDecimal.valueOf(rs.getInt("contract_number"));
        Action action = Action.valueOf(rs.getString("action_fk"));

        if (action.equals(Action.BUY)) {
            premium = premium.multiply(BigDecimal.valueOf(-1));
        }

        return TradeSummaryModel.builder()
                .withSymbol(rs.getString("symbol"))
                .withTotal(premium.multiply(quantity).multiply(new BigDecimal(100)).subtract(fees))
                .withTrades(1)
                .build();
    }

    private TradeSummaryModel fromDividend(ResultSet rs) throws SQLException {
        BigDecimal dividend = BigDecimal.valueOf(rs.getDouble("dividend"));
        BigDecimal quantity = BigDecimal.valueOf(rs.getInt("quantity"));

        return TradeSummaryModel.builder()
                .withSymbol(rs.getString("symbol"))
                .withTotal(dividend.multiply(quantity))
                .withTrades(1)
                .build();
    }
}

File: services/tradelog/tradelog-api/src/main/kotlin/com/example/tradelog/api/db/converter/TradeSummaryModelRowMapper.kt
package com.example.tradelog.api.db.converter

import com.example.tradelog.api.core.model.ActionType
import com.example.tradelog.api.core.model.TradeSummaryModel
import com.example.tradelog.api.core.model.TransactionType
import org.springframework.jdbc.core.RowMapper
import java.math.BigDecimal
import java.sql.ResultSet

class TradeSummaryModelRowMapper: RowMapper<TradeSummaryModel> {

    override fun mapRow(rs: ResultSet, rowNum: Int): TradeSummaryModel {

        return when(val transactionType = TransactionType.lookup(rs.getString("transaction_type_fk"))) {
            TransactionType.SHARE -> mapStock(rs)
            TransactionType.OPTION -> mapOption(rs)
            TransactionType.DIVIDEND -> mapDividend(rs)
            else -> throw IllegalArgumentException("Received unknown transaction type: $transactionType")
        }
    }

    private fun mapStock(rs: ResultSet): TradeSummaryModel {
        val actionType = ActionType.lookup(rs.getString("action_fk"))
        var price = BigDecimal.valueOf(rs.getDouble("price"))
        val fees = BigDecimal.valueOf(rs.getDouble("broker_fees"))
        val quantity = BigDecimal.valueOf(rs.getInt("quantity").toLong())

        if (ActionType.BUY === actionType) {
            price = price.negate()
        }
        return TradeSummaryModel(
                symbol = rs.getString("symbol"),
                trades = 1,
                total = price.multiply(quantity).subtract(fees),
                legClosed = rs.getBoolean("leg_closed")
        )
    }

    private fun mapOption(rs: ResultSet): TradeSummaryModel {
        val actionType = ActionType.lookup(rs.getString("action_fk"))
        var premium = BigDecimal.valueOf(rs.getDouble("premium"))
        val fees = BigDecimal.valueOf(rs.getDouble("broker_fees"))
        val quantity = BigDecimal.valueOf(rs.getInt("contract_number").toLong())

        if (ActionType.BUY === actionType) {
            premium = premium.negate()
        }

        return TradeSummaryModel(
                symbol = rs.getString("symbol"),
                trades = 1,
                total = premium.multiply(quantity).multiply(BigDecimal(100)).subtract(fees),
                legClosed = rs.getBoolean("leg_closed")
        )
    }

    private fun mapDividend(rs: ResultSet): TradeSummaryModel {
        val dividend = BigDecimal.valueOf(rs.getDouble("dividend"))
        val quantity = BigDecimal.valueOf(rs.getInt("quantity").toLong())

        return TradeSummaryModel(
                symbol = rs.getString("symbol"),
                trades = 1,
                total = dividend.multiply(quantity),
                legClosed = true
        )
    }
}
